modules/caddyhttp/reverseproxy/reverseproxy.go

```diff
@@ -383,24 +399,51 @@ func (h *Handler) Provision(ctx caddy.Context) error {
 // Cleanup cleans up the resources made by h.
 func (h *Handler) Cleanup() error {
 	// close hijacked connections (both to client and backend)
-	var err error
-	h.connectionsMu.Lock()
-	for _, oc := range h.connections {
-		if oc.gracefulClose != nil {
-			// this is potentially blocking while we have the lock on the connections
-			// map, but that should be OK since the server has in theory shut down
-			// and we are no longer using the connections map
-			gracefulErr := oc.gracefulClose()
-			if gracefulErr != nil && err == nil {
-				err = gracefulErr
+	closeConnections := func() error {
+		var err error
+		h.connectionsMu.Lock()
+		defer h.connectionsMu.Unlock()
+
+		for _, oc := range h.connections {
+			if oc.gracefulClose != nil {
+				// this is potentially blocking while we have the lock on the connections
+				// map, but that should be OK since the server has in theory shut down
+				// and we are no longer using the connections map
+				gracefulErr := oc.gracefulClose()
+				if gracefulErr != nil && err == nil {
+					err = gracefulErr
+				}
```

Comment:
Slightly nervous about this, but maybe we can revisit this one more time before merge.

Reply:
Me too, OK.